substrate/client/network/src/litep2p/shim/notification/tests/peerset.rs

```diff
@@ -0,0 +1,812 @@
+	to_peerset
+		.unbounded_send(PeersetCommand::AddReservedPeers { peers: new_reserved_peers.clone() })
+		.unwrap();
+
+	match peerset.next().await {
+		Some(PeersetNotificationCommand::OpenSubstream { peers: out_peers }) => {
+			assert_eq!(out_peers.len(), 2);
+			assert!(out_peers.iter().find(|peer| peer == &&common_peer).is_none());
+
+			for peer in &out_peers {
+				assert!(!reserved.contains(peer));
+
+				if peer != &common_peer {
+					assert!(peerset.reserved_peers().contains(peer));
+					assert_eq!(
+						peerset.peers().get(peer),
+						Some(&PeerState::Opening { direction: Direction::Outbound(Reserved::Yes) }),
+					);
+				}
+			}
+		},
+		event => panic!("invalid event: {event:?}"),
+	}
+
+	// verify the `common_peer` peer between the reserved sets is still in the state `Open`
+	assert_eq!(
+		peerset.peers().get(&common_peer),
+		Some(&PeerState::Connected { direction: Direction::Outbound(Reserved::Yes) })
+	);
+}
+
+#[tokio::test]
+async fn opening_peer_gets_canceled_and_disconnected() {
+	sp_tracing::try_init_simple();
+
+	let peerstore_handle = Arc::new(peerstore_handle_test());
+	let _reserved = (0..1)
+		.map(|_| {
+			let peer = PeerId::random();
+			peerstore_handle.add_known_peer(peer);
+			peer
+		})
+		.collect::<Vec<_>>();
+	let num_connected = Arc::new(Default::default());
+	let (mut peerset, to_peerset) = Peerset::new(
+		ProtocolName::from("/notif/1"),
+		25,
+		25,
+		false,
+		Default::default(),
+		Arc::clone(&num_connected),
+		peerstore_handle,
+	);
+	assert_eq!(peerset.num_in(), 0);
+	assert_eq!(peerset.num_out(), 0);
+
+	let peer = match peerset.next().await {
+		Some(PeersetNotificationCommand::OpenSubstream { peers: out_peers }) => {
+			assert_eq!(peerset.num_in(), 0);
+			assert_eq!(peerset.num_out(), 1);
+			assert_eq!(out_peers.len(), 1);
+
+			for peer in &out_peers {
+				assert_eq!(
+					peerset.peers().get(&peer),
+					Some(&PeerState::Opening { direction: Direction::Outbound(Reserved::No) })
+				);
+			}
+
+			out_peers[0]
+		},
+		event => panic!("invalid event: {event:?}"),
+	};
+
+	// disconnect the now-opening peer
+	to_peerset.unbounded_send(PeersetCommand::DisconnectPeer { peer }).unwrap();
+
+	// poll `Peerset` to register the command and verify the peer is now in state `Canceled`
+	futures::future::poll_fn(|cx| match peerset.poll_next_unpin(cx) {
+		Poll::Pending => Poll::Ready(()),
+		_ => panic!("unexpected event"),
+	})
+	.await;
+
+	assert_eq!(
+		peerset.peers().get(&peer),
+		Some(&PeerState::Canceled { direction: Direction::Outbound(Reserved::No) })
+	);
+	assert_eq!(peerset.num_out(), 1);
+
+	// report to `Peerset` that the substream was opened, verify that it gets closed
+	assert!(!peerset.report_substream_opened(peer, traits::Direction::Outbound));
+	assert_eq!(
+		peerset.peers().get(&peer),
+		Some(&PeerState::Closing { direction: Direction::Outbound(Reserved::No) })
+	);
+	assert_eq!(num_connected.load(Ordering::SeqCst), 1);
```

Comment:
As `num_connected` operations in `Peerset` are `Relaxed`, not sure if `SeqCst` operations in the tests work as desired. On the other hand, all the tests here are single-threaded, so this probably doesn't matter.

Reply:
It hasn't caused problems with tests so far so I'll keep it as-is and change the ordering if any issues pop up